### ISSUE TITLE
Video stream is earlier than audio stream

### ISSUE THREAD

**Issue by steeveen** [bytedeco/javacv#1722]
Hello，I have used JavaCV from several months. Recentlly, I wanna to use it to transfer stream from one rtmp source A to another rtmp
source B. but I find when I play B on Potplayer, Video stream is earlier than Audio stream in most scenarios. and the latency time is random. In rate case, Video Steam is sync with Audio Stream.
The information I have:
I use OBS to push a stream( from A local file ) to rtmp://A
Init _FFmpegFrameGrabber_  with _rtmp://A_ and _grabFrame(audio:true,video:true,processing:true,keyframe:false,data:false)_
Init _FFmpegFrameRecorder_  with _rtmp://B and record(Frame)
Play rtmp://B on Potplayer or VLC.
The key code：
`
  try (FFmpegFrameGrabber grabber = new FFmpegFrameGrabber(rtmpA);) {
            grabber.setOption("s_timeout", String.valueOf(10 * 1000_000));
            grabber.setOption("rw_timeout", String.valueOf(10 * 1000_000));
            grabber.start();

            try (FFmpegFrameRecorder recorder = new FFmpegFrameRecorder(rtmpB, grabber.getImageWidth(), grabber.getImageHeight(), grabber.getAudioChannels())) {
                recorder.setVideoOption("crf", "18");
//                recorder.setGopSize(2);
                recorder.setFrameRate(grabber.getVideoFrameRate());
                recorder.setVideoCodec(AV_CODEC_ID_H264);
                recorder.setFormat("flv");
                recorder.setMaxDelay(0);
                recorder.setInterleaved(true);
                recorder.setVideoOption("tune", "zerolatency");
                recorder.setVideoOption("preset", "ultrafast");
                recorder.setVideoOption("rtsp_transport", "tcp");`
                recorder.start();
                Frame frame;
                while ((frame = grabber.grabFrame(true,true,true,false,false)) != null) {
                    slideWindow.add(frame.clone());
                }
`
The method I have Tried:
have tried many combinations of parameters。such as recorder.setTimestamp(grabber.getTimestamp)，recorder.setOption("vsync",xx).  recorder.setFrameRate(a fixed number)but none of them worked.
For the normal run of this basic function, I'm about to crash. Looking for your response.

**Comment by saudet** [comment]
Maybe your CPU isn't able to encode in real time...

**Comment by steeveen** [comment]
Emmm I’m afried not. The cpu I tested is Intel.
A fact is that I use OBS to transfer source A to source B on the same machine and it works.

**Comment by saudet** [comment]
Maybe OBS is using libx264 but you're not?

**Comment by steeveen** [comment]
Oh. A nice reminder.
What should I do in the code if I want to use libx264

**Comment by saudet** [comment]
Make sure to use the GPL enabled binaries:
https://github.com/bytedeco/javacv/blob/master/samples/pom.xml#L24

**Comment by steeveen** [comment]
Hello saudet.
I have tried according to your advise. I replaced ffmpeg-pllatform with ffmpeg-platform-gpl. But sadly it change nothing....
The CPU I used is Intel i7-8700.  The JavaCV version I tested on is 1.5.5.
I use ffmpeg like:
> ffmpeg -i rtmp://A -vcodec copy -acodec copy -f flv rtmp://B
And this can work normally.  So the normally runnning of OBS and FFmpeg can show that my hardware can support it.
Besides. I also tried theses and grouped them by result:
sync:
1. mp4 file -----JavaCV----->rtmp 
2. mp4 link -----JavaCV----->rtmp
3. mp4 link-----OBS----->rtmp-----JavaCV----->rtmp
4. mp4 link-----OBS----->rtmp-----JavaCV----->mp4 file
5. mp4 file-----OBS----->rtmp-----OBS----->rtmp
not sync:
6. mp4 file -----OBS----->rtmp  -----JavaCV-----> mp4 file
7. mp4 file -----OBS----->rtmp  -----JavaCV----->rtmp

I think that OBS change something from mp4 file which may can't be recognized by JavaCV and result in 7 wrong. But according to 5,the change made by OBS can be recognized by itself.

I set my OBS params as: 

- videoBitRate:same as mp4 file.(9041Kbps)
- video encoder:software(x264)
- audioBitRate:same as mp4 file(320)
- enable advanced encoder setting:false
- frameRate:same as mp4 file (25)
- sampleRate:same as mp4 file(48kHz)

Does anyone have idea about my guess...

**Comment by saudet** [comment]
> Hello saudet. I have tried according to your advise. I replaced ffmpeg-pllatform with ffmpeg-platform-gpl. But sadly it change nothing.... The CPU I used is Intel i7-8700. The JavaCV version I tested on is 1.5.5.

Please try again with JavaCV 1.5.6.
 
> I used ffmpeg like:
> 
> > ffmpeg -i rtmp://A -vcodec copy -acodec copy -f flv rtmp://B
> 
> And this can work normally. So the normally runnning of OBS and FFmpeg can show that my hardware can support it.

Please try again with the version of ffmpeg that comes with JavaCV to make sure its binaries actually support your hardware.

**Comment by saudet** [comment]
BTW, if you're not tying to transcode, but just want to forward (copy) the packets, please use grabPacket() and recordPacket() instead of grabFrame() and record(Frame).

**Comment by steeveen** [comment]
My requirement is do something on video frame (Such as flip、beauty or brighten) before it is pushed into recorder. So I have to decode it.
Anyway, I will have a try on the specific JavaCV and ffmpeg.

**Comment by saudet** [comment]
We can do those things with filters and the ffmpeg program too...

**Comment by steeveen** [comment]
Hi saudet. I come back again.
As I expected. I upgraded JavaCV to 1.5.6(ffmpeg-platform 4.4-1.5.6, opencv-platform 4.5.3-1.5.6) but it changed nothing.
I have hearded of ffmpeg filters but I want to customize the filter function. For example,  I want to process the pixel matrix(Frame's essence in Math) by my brighten algorithm. So I have to decode the frame data. Can FFmpegFrameFilter give me a hand on this?

**Comment by saudet** [comment]
If you want to process images in Java, it won't work with the ffmpeg program, correct. But for starters, you should make sure that the ffmpeg program *can* work with a couple of filters that you may want to use anyway. If that's not able to function correctly, JavaCV probably won't either.

**Comment by steeveen** [comment]
I just used FFmpegFrameFilter to crop as a Demo. the filter took effect. But  filters still has the same problem that audio stream is  behind video stream.
I can convert the data between Frame and BufferedImage with your Java2DFrameConverter, which can make my previous skills come in handy(OpenCV or other tools...). With the converter I can implement all I need.
But the key point is still that stream processed by my code is out of sync. Even I don't do anything just:
`recorder.record(grabber.grabFrame(true,true,true,false,false)`

**Comment by saudet** [comment]
No, I mean, if you do some filtering with the ffmpeg *program*, not with JavaCV, do you have any problems with that?

**Comment by steeveen** [comment]
O. I use crop filter with ffmpeg cmd and..... all things run normally.

**Comment by steeveen** [comment]
I don't have environment to test on ffmpeg c program.....

**Comment by saudet** [comment]
Like I said, please try with the version of ffmpeg that comes with JavaCV to make sure its binaries actually support your hardware:
http://bytedeco.org/javacpp-presets/ffmpeg/apidocs/org/bytedeco/ffmpeg/ffmpeg.html

**Comment by steeveen** [comment]
Yes. Actually I had tried this .....
My code:

```
String ffmpeg=Loader.load(org.bytedeco.ffmpeg.ffmpeg.class);
ProcessBuilder pb=new ProcessBuilder(ffmpeg,"-i","rtmp://A","-f","flv","-vcodec","copy","-acodec","copy","rtmp://B");
pb.inheritIO().start().waitFor();
```
And it runs in sync exactly....

**Comment by saudet** [comment]
That doesn't do any filtering. Like I said, use grabPacket() and
recordPacket() for that.

**Comment by steeveen** [comment]
`recorder.recordPacket(grabber.grabPacket())` runs in sync

**Comment by steeveen** [comment]
Things looks like happening in unpacketing?

**Comment by saudet** [comment]
Does it happen only with RTMP? Or does it also happen with files?

**Comment by steeveen** [comment]
Yes. It is just happen from rtmp to rtmp.
from files to rtmp or from rtmp to file both work in sync.

**Comment by saudet** [comment]
This may have been fixed with FFmpeg 5.0.
Please give it try with the snapshots: http://bytedeco.org/builds/

@bradh Any ideas?

**Comment by bradh** [comment]
Sorry, no suggestions.
